dlib/unicode/unicode.h

```diff
@@ -493,23 +499,63 @@ namespace dlib
         invalid_utf8_error():error(EUTF8_TO_UTF32) {}
     };
 
-    inline const ustring convert_utf8_to_utf32 (
-        const std::string& str
+    template <typename forward_iterator, typename unary_op>
+    inline void convert_utf8_to_utf32(
+        forward_iterator ibegin,
+        forward_iterator iend,
+        unary_op op
     )
     {
-        using namespace unicode_helpers;
-        ustring temp;
-        std::istringstream sin(str);
 
-        temp.reserve(str.size());
+        using char_type = std::decay_t<decltype(*ibegin)>;
+        static_assert(std::is_same<char_type, char>::value ||
+                      std::is_same<char_type, wchar_t>::value ||
+                      std::is_same<char_type, unichar>::value,
+                      "char_type must be either char or unichar");
+
+        if (std::is_same<char_type, unichar>::value)
+        {
+            while (ibegin != iend)
+                op(*(ibegin++));
+            return;
+        }
+
+        if (std::is_same<char_type, wchar_t>::value)
+        {
+            return;
+        }
```

Comment:
Here, we should handle this case properly, right?
Or there's another way?

Reply:
Ha yeah totally have to handle this somehow.

Reply:
Oh look at `convert_wstring_to_utf32()`.  That's got the `wchar_t` to `unichar` conversion.  Stick that in here and then this can just be the one glorious `convert_to_utf32()` function that converts any character sequence to utf32.

Reply:
I will have to make that function also take a range and a lambda so that it doesn't allocate, just like the `u8_to_u32`.